domain/resource/state/resource.go

```diff
@@ -1356,6 +1481,542 @@ WHERE uuid = $updatePotentialResource.uuid
+		}
+
+		unitResourcesToSet = append(unitResourcesToSet, unitResource{
+			ResourceUUID: resUUID.String(),
+			UnitUUID:     unitUUID.String(),
+			AddedAt:      res.Timestamp,
+		})
+	}
+	return unitResourcesToSet, nil
+}
+
+// insertResources inserts resources, application resources, kubernetes
+// resources and unit resources using bulk inserts.
+func (st *State) insertResources(
+	ctx context.Context,
+	tx *sqlair.TX,
+	toSet *resourcesToSet,
+) error {
+	// Bulk insert the resources.
+	if len(toSet.resources) > 0 {
+		insertStmt, err := st.Prepare(`
+INSERT INTO resource (*) VALUES ($setResource.*)
+`, setResource{})
+		if err != nil {
+			return errors.Capture(err)
+		}
+		err = tx.Query(ctx, insertStmt, toSet.resources).Run()
+		if err != nil {
+			return errors.Errorf("inserting resources: %w", err)
+		}
+	}
+
+	// Bulk insert the application-resource links.
+	if len(toSet.applicationResources) > 0 {
+		insertApplicationResourceStmt, err := st.Prepare(`
+INSERT INTO application_resource (*) VALUES ($applicationResource.*)
+`, applicationResource{})
+		if err != nil {
+			return errors.Capture(err)
+		}
+		err = tx.Query(ctx, insertApplicationResourceStmt, toSet.applicationResources).Run()
+		if err != nil {
+			return errors.Errorf("linking resources to applications: %w", err)
+		}
+	}
+
+	// Bulk insert the kubernetes-application-resource links.
+	if len(toSet.kubernetesApplicationResources) > 0 {
+		insertK8sResourceStmt, err := st.Prepare(`
+INSERT INTO kubernetes_application_resource (*) VALUES ($kubernetesApplicationResource.*)
+`, kubernetesApplicationResource{})
+		if err != nil {
+			return errors.Capture(err)
+		}
+		err = tx.Query(ctx, insertK8sResourceStmt, toSet.kubernetesApplicationResources).Run()
+		if err != nil {
+			return errors.Errorf("linking resources to kubernetes applications: %w", err)
+		}
+	}
+
+	// Bulk insert the unit-resource links.
+	if len(toSet.unitResources) > 0 {
+		insertUnitResourceStmt, err := st.Prepare(`
+INSERT INTO unit_resource (*) VALUES ($unitResource.*)
+`, unitResource{})
+		if err != nil {
+			return errors.Capture(err)
+		}
+		err = tx.Query(ctx, insertUnitResourceStmt, toSet.unitResources).Run()
+		if err != nil {
+			return errors.Errorf("linking resources to units: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// isLocalCharm returns true if the charm uuid belongs to a local charm.
+func (st *State) isLocalCharm(
+	ctx context.Context,
+	tx *sqlair.TX,
+	charmID charm.ID,
+) (bool, error) {
+	source := getCharmSource{
+		UUID: charmID.String(),
+	}
+	getCharmSourceStmt, err := st.Prepare(`
+SELECT cs.name AS &getCharmSource.source_name
+FROM   charm c
+JOIN   charm_source cs ON c.source_id = cs.id
+AND    uuid = $getCharmSource.uuid
+`, source)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, getCharmSourceStmt, source).Get(&source)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	return charm.Local.Matches(source.SourceName), nil
```

Comment:
This works by accident. You shouldn't be using the core charm package in state. The values in core charm package are either `local` or `ch`, yet in the database, they're `local` or `charmhub`. I did this on purpose.

https://github.com/juju/juju/blob/8111f59ad7af3eabacf4a0b49e251cb1fc187ced/domain/schema/model/sql/0015-charm.sql#L201-L202

This is incorrect.

You just need to check if the `COUNT(*)`. Drop the local charm check.

Reply:
@Aflynn50 to test, do a migration of local and repository charms, then trigger the charm revision updater worker via its wrench.